modules/common/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/EnrichmentManager.scala

```diff
@@ -68,44 +68,114 @@ object EnrichmentManager {
     featureFlags: EtlPipeline.FeatureFlags,
     invalidCount: F[Unit],
     registryLookup: RegistryLookup[F],
-    atomicFields: AtomicFields
-  ): EitherT[F, BadRow, EnrichedEvent] =
-    for {
-      enriched <- EitherT.rightT[F, BadRow](new EnrichedEvent)
-      extractResult <- mapAndValidateInput(
-                         raw,
-                         enriched,
-                         etlTstamp,
-                         processor,
-                         client,
-                         registryLookup
-                       )
-      _ = {
-        ME.formatUnstructEvent(extractResult.unstructEvent).foreach(e => enriched.unstruct_event = e)
-        ME.formatContexts(extractResult.contexts).foreach(c => enriched.contexts = c)
-      }
-      enrichmentsContexts <- runEnrichments(
-                               registry,
-                               processor,
-                               raw,
-                               enriched,
-                               extractResult.contexts,
-                               extractResult.unstructEvent,
-                               featureFlags.legacyEnrichmentOrder
-                             )
-      _ = ME.formatContexts(enrichmentsContexts ::: extractResult.validationInfoContexts).foreach(c => enriched.derived_contexts = c)
-      _ <- validateEnriched(
-             enriched,
-             raw,
-             enrichmentsContexts,
-             client,
-             processor,
-             registryLookup,
-             featureFlags.acceptInvalid,
-             invalidCount,
-             atomicFields
-           )
-    } yield enriched
+    atomicFields: AtomicFields,
+    emitIncomplete: Boolean
+  ): IorT[F, BadRow, EnrichedEvent] = {
+    def enrich(enriched: EnrichedEvent): IorT[F, NonEmptyList[NonEmptyList[Failure]], List[SelfDescribingData[Json]]] =
+      for {
+        extractResult <- mapAndValidateInput(
+                           raw,
+                           enriched,
+                           etlTstamp,
+                           processor,
+                           client,
+                           registryLookup
+                         )
+                           .leftMap(NonEmptyList.one)
+                           .possiblyExitingEarly(emitIncomplete)
+        // Next 2 lines remove the invalid contexts and the invalid unstructured event from the event.
+        // This should be done after the bad row was created and only if emitIncomplete is enabled.
```

Comment:
I think this comment should be removed because next two lines should be performed in all cases not just when `emitIncomplete` is enabled because `contexts` and `unstruct_event` might need to be updated if there is self describing json with superseded schema.